src/kafka-net/BrokerRouter.cs

```diff
@@ -121,30 +123,60 @@ public List<Topic> GetTopicMetadata(params string[] topics)
         /// <param name="topics">List of topics to update metadata for.</param>
         /// <remarks>
         /// This method will ignore the cache and initiate a call to the kafka servers for all given topics, updating the cache with the resulting metadata.
-        /// Only call this method to force a metadata update.  For all other queries use <see cref="GetTopicMetadata"/> which uses cached values.
+        /// Only call this method to force a metadata update.  For all other queries use <see cref="GetTopicMetadataFromLocalCache"/> which uses cached values.
         /// </remarks>
-        public void RefreshTopicMetadata(params string[] topics)
+
+        public Task<bool> RefreshTopicMetadata(params string[] topics)
         {
-            //TODO need to remove lock here, try and move to lock free design
-            lock (_threadLock)
+            return RefreshTopicMetadata(_kafkaOptions.CacheExpiration, _kafkaOptions.RefreshMetadataTimeout, topics);
+        }
+
+        /// <summary>
+        /// refresh metadata Request get to server for all topic that there Cache Expire.
+        /// CacheExpiration is max time for topic to be valid after this time the Cache Expire for topic,
+        /// and be refresh on next refresh metadata Request.
+        /// if cacheExpiration is null: refresh metadata Request get to server for all topic that dose not exists on Cache.
+        /// </summary>
+        private async Task<bool> RefreshTopicMetadata(TimeSpan? cacheExpiration, TimeSpan timeout, params string[] topics)
+        {
+            try
             {
-                _kafkaOptions.Log.DebugFormat("BrokerRouter: Refreshing metadata for topics: {0}", string.Join(",", topics));
+                await _taskLocker.WaitAsync(timeout);
+                int missingFromCache = SearchCacheForTopics(topics, cacheExpiration).Missing.Count;
+                if (missingFromCache == 0)
+                {
+                    return false;
+                }
 
+                _kafkaOptions.Log.DebugFormat("BrokerRouter: Refreshing metadata for topics: {0}", string.Join(",", topics));
                 //get the connections to query against and get metadata
                 var connections = _defaultConnectionIndex.Values.Union(_brokerConnectionIndex.Values).ToArray();
-                var metadataResponse = _kafkaMetadataProvider.Get(connections, topics);
+                var taskMetadata = _kafkaMetadataProvider.Get(connections, topics);
+                await Task.WhenAny(Task.Delay(timeout), taskMetadata);
+                if (!taskMetadata.IsCompleted)
+                {
+                    var ex = new Exception("_kafkaMetadataProvider.Get not get in time");
```

Comment:
Exception error messages should be readable, for example: "Metadata refresh operation timed out"